### PR TITLE
DPE-4643 Limit username to 32chars

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

New server format for username can lead to bigger than 32 chars.

## Solution

Ensure generated username is not bigger than 32 chars.

### CODE REVIEW

**Review by shayancanonical — APPROVED** [review]

File: src/relations/database_provides.py [hunk 1]
```diff
@@ -50,11 +50,12 @@ def __eq__(self, other) -> bool:
...
     def _get_username(self, database_requires_username: str) -> str:
         """Database username"""
-        # Prefix username with username from database requires relation.
+        # Prefix username with username from database requires relation truncating to the
+        # final 32 chars (MySQL username limit)
         # This ensures a unique username if MySQL Router is deployed in a different Juju model
         # from MySQL.
         # (Relation IDs are only unique within a Juju model.)
-        return f"{database_requires_username}-{self._id}"
+        return f"{database_requires_username}-{self._id}"[-32:]
```
[the remark this answers is not in the document]
Comment (taurus-forever) [reply src/relations/database_provides.py:58]: Agree to Carl here. id is just an integer... I would honestly just take md5({database_requires_username}-{self._id}) so, it always be 32 symbols.... and will  not allow to brootforcing usernames by increasing id+1.

**Review by taurus-forever — APPROVED** [review]
LGTM, but I believe Carl have a point.

File: src/relations/database_provides.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (taurus-forever) [reply src/relations/database_provides.py:58]: You will have no relation ID anyway. :-(
Commented in https://github.com/canonical/mysql-k8s-operator/pull/439#pullrequestreview-2135360246

**Comment by paulomach** [comment]
Closing as solution is server side only

**Pull request closed without merging**